File: src/data_structures/07_heap/implementations/rs/heap.rs
pub fn heap(array: &mut Vec<i32>) {
  let heap_size = array.len();

  for i in (0..heap_size / 2).rev() {
    heapify(array, i, heap_size);
  }
}

pub fn heapify(heap: &mut Vec<i32>, i: usize, heap_size: usize) {
  let mut largest = i;
  let left = 2 * i + 1;
  let right = 2 * i + 2;

  if left < heap_size && heap[left] > heap[largest] {
    largest = left;
  }

  if right < heap_size && heap[right] > heap[largest] {
    largest = right;
  }

  if largest != i {
    heap.swap(i, largest);
    heapify(heap, largest, heap_size);
  }
}

pub fn parent(index: usize) -> usize {
  (index - 1) / 2 // Implicitly rounded down
}

pub fn insert(heap: &mut Vec<i32>, value: i32) {
  heap.push(value);
  let mut index = heap.len() - 1;

  while index > 0 && heap[parent(index)] < heap[index] {
    heap.swap(parent(index), index);
    index = parent(index)
  }
}

pub fn extract(heap: &mut Vec<i32>) -> Option<i32> {
  if heap.len() == 0 {
    return None;
  } else {
    let max_val = heap[0];
    heap[0] = heap[heap.len() - 1];
    heap.pop();
    heapify(heap, 0, heap.len());

    Some(max_val)
  }
}

File: src/data_structures/07_heap/implementations/go/heap.go
package heap

import "errors"

func Heap(array []int) {
	for i := len(array) / 2; i >= 0; i-- {
		Heapify(array, i, len(array))
	}
}

func Heapify(heap []int, i int, heapSize int) {
	largest := i
	left := 2*i + 1
	right := 2*i + 2

	if left < heapSize && heap[left] > heap[largest] {
		largest = left
	}
	if right < heapSize && heap[right] > heap[largest] {
		largest = right
	}
	if largest != i {
		heap[i], heap[largest] = heap[largest], heap[i]

		Heapify(heap, largest, heapSize)
	}
}

func Parent(index int) int {
	return (index - 1) / 2
}

func Insert(heap *[]int, value int) {
	*heap = append(*heap, value) // Append the new value to the heap
	index := len(*heap) - 1      // Get the index of the last element

	// Bubble up the new value to restore heap property
	for index > 0 && (*heap)[Parent(index)] < (*heap)[index] {
		// Swap with parent if the current value is larger
		(*heap)[Parent(index)], (*heap)[index] = (*heap)[index], (*heap)[Parent(index)]
		index = Parent(index) // Move up the heap
	}
}

func Extract(heap *[]int) (int, error) {
	if len(*heap) == 0 {
		return 0, errors.New("Heap is empty")
	}
	maxVal := (*heap)[0]
	(*heap)[0] = (*heap)[len(*heap)-1]
	*heap = (*heap)[:len(*heap)-1]
	Heapify(*heap, 0, len(*heap))

	return maxVal, nil
}
